common/src/main/java/com/wynntils/features/chat/ChatMentionFeature.java

```diff
@@ -61,69 +61,44 @@ private Pattern buildPattern() {
+        StyledText styledText = StyledText.fromComponent(message);
 
-        if (looseMatcher.find()) {
-            if (markMention.get()) {
-                e.setMessage(rewriteComponentTree(message));
-            }
-            if (dingMention.get()) {
-                McUtils.playSoundUI(SoundEvents.NOTE_BLOCK_PLING.value());
-            }
-        }
-    }
-
-    private Component rewriteComponentTree(Component comp) {
-        // Make a copy of the component without the siblings
-        MutableComponent curr = MutableComponent.create(comp.getContents()).withStyle(comp.getStyle());
-        // .getString() is used here as it gives formattingchars when those exist. It is needed for guild messages
-        // because wynn still uses legacy coloring for it.
-        String text = curr.getString();
-
-        // if current component has text check it for mentions and rewrite the component if necessary
-        if (!text.isEmpty()) {
-            curr = rewriteMentions(curr, text, comp.getStyle());
-        }
+        StyledTextPart partFinding = styledText.getPartFinding(mentionPattern, PartStyle.StyleType.NONE);
 
-        // process and append siblings
-        for (Component c : comp.getSiblings()) {
-            curr.append(rewriteComponentTree(c));
+        // If the part is null, it means that we did not get mentioned,
+        // or the mention is in two different parts, we likely don't want to mark it.
+        if (partFinding == null) {
+            return;
         }
 
-        return curr;
-    }
-
-    private MutableComponent rewriteMentions(MutableComponent curr, String text, Style style) {
-        Matcher match = mentionPattern.matcher(text);
-
-        // if we match then rewrite the component if there are no matches the function will just return
-        if (match.find()) {
-            // get the start of the string before any mention and set curr to be it
-            curr = Component.literal(text.substring(0, match.start())).withStyle(style);
-
-            // do the name of the first mention
-            curr.append(Component.literal(mentionColor.get() + match.group(0)));
+        if (markMention.get()) {
+            Matcher matcher = styledText.getMatcher(mentionPattern, PartStyle.StyleType.NONE);
 
-            // Store the point at which this match ended
-            int lastEnd = match.end();
+            if (!matcher.find()) {
+                WynntilsMod.error(
+                        "Found a part matching the mention pattern, but couldn't find a match in the string.");
+                return;
+            }
 
-            // Handle any additional mentions within the message
-            while (match.find()) {
-                // get the bit in between of matches
-                curr.append(Component.literal(text.substring(lastEnd, match.start())))
-                        .withStyle(style);
+            StyledText newText = styledText.splitAt(matcher.start()).splitAt(matcher.end());
```

Comment:
This logic seems frail, and not general enough. Getting matcher start position seems scary and probably incorrect -- that matching is done using the coded string, so matching "§4b" for b will give position 2, even though the entire string is just 1 character long.  You will also fail to highlight multiple mentions in the same message. Perhaps not strictly needed, but does look like a bug.

Reply:
Instead, I think you should have a iterating callback interface to StyledText, eg. like this:
```
styledText.process(part -> {
if (matches(part, pattern)) {
List<Part> splitted = part.splitInThreeAndHighlighMatch(part, patttern);
return splitted;
} else {
return null; // no change to part
}})
```

and let the callback to process return either `null` == no change to the part, a new part == replace it with the new part, List<Part> == replace it with this list of parts (don't forget to check for empty parts!), or finally `StyledTextPart.EMPTY` (a new constant with an empty string), which means we want to delete this part.

Reply:
(an alternative implementation is to just return the unchanged part if it should not be changed, and return null if it should be deleted)

Reply:
That also  mean I think we should remove the `splitAt`, `getPartMatching` and `replacePart` functions on StyledText, since I think they are likely to get used in a way that introduce bugs, i.e. they are not helpful to us for writing correct code.

Reply:
But this 3 way split is needed for multiple features. Should I make it an util?